### ISSUE TITLE
RFC: EventTimer abstraction to support libproc simulations

### ISSUE THREAD

**Issue by lucbouchard1** [PolySat/libproc#12]
# RFC: Event Timer Abstraction

We would like to add support for doing accelerated simulations of libproc processes, as this would enable simulating the PolySat ADCS. To accomplish this, we need allow for virtualization of libproc's timing system, which means that timed events are immediately popped of the queue and executed with no delay. An integer stores the current virtual time, and the value is incremented whenever a timed event is executed with the associated time.

To implement virtual time, this issue presents the `EventTimer` interface, which is a set of functions called during the libproc event loop to control timing. Implementations of this interface can be swapped out to allow for alternate event loop flows. An example implementation is available in #8; however, here I present alternate API's and request commentary. Many of these API options were discussed in the comments on #8, but they are formalized here.

## Option 1: `select` Interface
This is what is currently implemented in #8, but it has some issues. The interface consists of the following functions:

``` C
   // Note the last params of the function are the same as the params to 'select'
   int (*block)(struct EventTimer *et, struct timeval *nextAwake,
          int nfds, fd_set *readfds, fd_set *writefds, fd_set *exceptfds);
   int (*get_gmt_time)(struct EventTimer *et, struct timeval *tv);
   int (*get_monotonic_time)(struct EventTimer *et, struct timeval *tv);
   void (*cleanup)(struct EventTimer *et);
```

And the flow within `EVT_start_loop` is as follows:

- Find next timed event on the queue and create `select` parameters using FD events.
- Call the EventTimer `block` function, passing the amount of time to the next timed event and all the select parameters. The `block` function calls `select` with the passed file descriptors, and blocks for however long is appropriate. For examples, see `eventTimer.c` in #8.
- Call callback functions of all events that have occurred.

One of these interfaces would exist for normal (default) timing, and one would exist for virtual timing. The virtual time `block` function would never block if there are timed events on the queue, it would just increment the virtual time, call select with `0` for the block time, then return. 

This interface provides function pointers to get the monotonic time and gmt time. When a process requests the time using `EVT_get_gmt_time` or `EVT_get_monotonic_time`, these function pointers are called instead of the standard `gettimeofday` to allow for time virtualization. Note that monotonic and gmt time are the same for virtual time.

### Pros:
- Simple to implement and comprehend, only consists of a few functions. 

### Cons:
- Forces EventTimer to use `select` over `poll`, `epoll`, etc. The idea of using the select parameters for the `block` function is pretty inflexible 

## Option 2: `EVTHandler` Interface
This implementation is identical to Option 1 except for one thing, the `block` function only takes two arguments: an `EventTimer` pointer and an `EVTHandler` pointer. Almost all the work of the event loop would be done in the `block` function.

### Pros:
- Since all the work is passed to the `EventTimer`, we could have multiple event loop implementations that rely on whichever system call (`select`, `poll`, `epoll`, whatever). The system call that is used could be swapped out at runtime depending on the requirements of the process. 
- EventTimer has access to all the process data, which means we could be more creative with what the event loop does (see end of post for examples)

### Cons:
- More boilerplate in the `EventTimer` implementations, leading to the possibility of bugs.
- Much harder to implement EventTimer's than Option 1

## Option 3: Multiple Function Call Interface
This implementation adds more functions to the Option 1 interface to simplify the implementation and remove dependence on `select`. It would look like the following:

``` C
   int (*block_time)(struct EventTimer *et, struct timeval *nextAwake, struct timeval *blockTime);
   int (*finish_block)(struct EventTimer *et);
   int (*get_gmt_time)(struct EventTimer *et, struct timeval *tv);
   int (*get_monotonic_time)(struct EventTimer *et, struct timeval *tv);
   void (*cleanup)(struct EventTimer *et);
```
And the flow within `EVT_start_loop` is as follows:

- Find next timed event on the queue.
- Call the EventTimer `block_time` function, passing the amount of time to the next timed event. The function populates the `blockTime` timeval with with amount of time the event loop should block. This will always be zero for virtual time if there are events on the queue.
- Call some blocking system call to detect FD events (`select`, `poll`, `epoll`, etc.)
- Call the finish_block function. Cleanups could happen here.
- Call callback functions of all events that have occurred.

### Pros:
- `EventTimer` does not have to worry about calling `select` or similar system calls, it's just concerned with time.
- The event loop can be reimplemented to use any system call relatively easily. It is not tied to select as in Option 1.

### Cons:
- The event loop must be rewritten to use a new system call, and we can't swap out implementations at runtime as in Option 2.
- The event timer doesn't have access to the process information. meaning we can't be as creative (see next section)

## `EventTimer` Creativity
One thing I realized while implementing these mid-event-loop call-out functions is that it allows us to extend the event loop quite a bit. For example, I implemented a libproc-debugger (code is on my computer) that broadcasts the current event queue contents over a socket, allowing clients to read the data and show the current status of the event queue. I wrote a simple python program that serves the data to a web page. Since the EventTimer controls the loop's timing, you can imagine adding pausing and stepping to the debugger. I didn't get around to implementing that.

This type of inventiveness is enabled by giving the EventTimer access to all the process data. Thus, I think we should keep this is consideration. One option to preserve this felxibility is to pass the EVTHanlder struct to more of the function calls considered in this issue, without the expectation that the user will use the data. Thus, we could imagine the EventTimer abstraction as a type of extension API.

**Comment by Joshua-Anderson** [comment]
Since we're probably not going to be implementing a new event timer unless we have a very specific use case, my 2 cents would be to go with option 2:

- This allows us to do almost anything with an event timer, so we don't have to anticipate new use cases.
- It does have more boiler plate, but because I don't see new event timers being created very frequently, I don't see this as a major issue.

That being said, I don't see any major issues with any of the choices, I think they all would work fine.

**Comment by LMPS97** [comment]
I don't think we should implement option 1. I feel it is very limited and that it acts as a bodge. However, since it is easy to implement, it might be our only option for EXOCube due to the lack of time, if it needs this code.

I like option 2. It allows us a lot of flexibility and potential future improvements. However, it is a lot of work and I don't know if we have the time for it if EXOCube needs this code. Thinking long forward, this would be the best option.

Option 3 is like the middle ground. I like its flexibility and easiness to implement it. But, as you said, we can't creative with it. 

Following josh's philosophy, my 2 cents are split in half between option 2 and 3. 

However, that been said, I don't have a lot of experience with libproc's internals compared to both of you.

**Comment by LMPS97** [comment]
Also, since Luc already implemented option 1, it should not be hard to modify it to get option 3.

**Comment by johnbellardo** [comment]
Resolved with PR #8
